pkg/compiler/consts.go

```diff
@@ -247,16 +271,28 @@ func (comp *compiler) patchConsts(consts0 map[string]uint64) {
 				}
 			case *ast.Call:
 				for _, attr := range n.Attrs {
-					if callAttrs[attr.Ident].HasArg {
+					if callAttrs[attr.Ident].Type == intAttr {
 						comp.patchTypeConst(attr.Args[0], consts, &missing)
 					}
 				}
 			case *ast.Struct:
 				for _, attr := range n.Attrs {
-					if structOrUnionAttrs(n)[attr.Ident].HasArg {
+					attrDesc := structOrUnionAttrs(n)[attr.Ident]
+					if attrDesc.Type == intAttr {
 						comp.patchTypeConst(attr.Args[0], consts, &missing)
 					}
 				}
+				foreachFieldAttr(n, func(attr *ast.Type, desc *attrDesc) {
```

Comment:
Is it possible to use the generic walking to reach all expression leafs instead of these special foreachFieldAttr/ForeachLeafType?
Walking is intended to reach all Type's that are present in the tree. So possible we don't need to do anything special here if we fix walking.

Reply:
I've replaced `ForeachLeafType` with `ast.Recursive` (I had to make `ast.Recursive`'s callback return a bool to control whether we go into the ast subtree).

`foreachFieldAttr` is now `foreachFieldAttrConst`, which is just a helper function. I don't think we need generic walking in there.